[dv/sram] flesh out DV documentation (#6678)
lowRISC/opentitan: hw/ip/sram_ctrl/doc/dv/index.md

* fatal_intg_error - signals a transmission integrity error
* fatal_parity_error - signals an internal parity error in the memory primitive

The alert agents provide the ability to independently drive and monitor both alert handshakes.

### SRAM_CTRL_LC_IF
The SRAM_CTRL IP has a simple sideband interface to the LC_CTRL to receive escalation requests.
This interface initializes the escalation bus and utilizes a simple task to drive escalation requests.

### SRAM_CTRL_EXEC_IF
The SRAM_CTRL IP has an interface to enable instruction execution from SRAM - allowing the Icache to fetch instruction data from the SRAM for the CPU.
This interface contains the necessary Lifecycle and OTP structs to enable and disable this functionality.

### UVM RAL Model
The SRAM_CTRL RAL model is created with the [`ralgen`]({{< relref "hw/dv/tools/ralgen/README.md" >}}) FuseSoC generator script automatically when the simulation is at the build stage.

It can be created manually by invoking [`regtool`]({{< relref "util/reggen/README.md" >}}):

<!-- TODO add description of second RAL model once implemented -->

### Reference models

> The SRAM_CTRL testbench uses a SystemVerilog golden model of its address and data [scrambling]({{< relref "hw/dv/sv/mem_bkdr_if" >}}) to check that encryption and decryption are performed correctly.

Reviewer: Just a question, I do not know if it is correct. `mem_bkdr_if` sounds more like a memory backdoor interface rather a reference models. But I do understand the SRAM encrypt/decrypt methods are checked inside that interface...

Author: hmm yes that's a good point. I should probably explain briefly that the golden model is within the mem_bkdr_if.

// a LC escalation request needs 3 cycles to be fully propagated through the DUT
parameter int LC_ESCALATION_PROPAGATION_CYCLES = 3;

typedef virtual mem_bkdr_if #(.MEM_PARITY(1)) mem_bkdr_vif;

typedef enum bit {
  SramCtrlRenewScrKey = 0,
  SramCtrlInit        = 1
} sram_ctrl_e;

typedef enum bit [1:0] {
  SramCtrlError           = 0,
  SramCtrlEscalated       = 1,
  SramCtrlScrKeyValid     = 2,
  SramCtrlScrKeySeedValid = 3
} sram_ctrl_status_e;
```
### TL_agent
SRAM_CTRL testbench instantiates (already handled in CIP base env) [tl_agent]({{< relref "hw/dv/sv/tl_agent/README.md" >}})
which provides the ability to drive and independently monitor random traffic via
TL host interface into SRAM_CTRL device.

### Alert agents
The SRAM_CTRL testbench instantiates 2 [alert agents]({{< relref "hw/dv/sv/alert_esc_agent/README.md" >}}) for:
* fatal_intg_error - signals a transmission integrity error
* fatal_parity_error - signals an internal parity error in the memory primitive

The alert agents provide the ability to independently drive and monitor both
alert handshakes.

### SRAM_CTRL_LC_IF
The SRAM_CTRL IP has a simple sideband interface to the LC_CTRL to receive escalation requests.

> This interface initializes the escalation bus and utilizes a simple task to

Reviewer 2: I thought the recommendation is to have sentences in a single line.

Author: thanks for catching that - this was most likely due to my editor auto-formatting line length...will fix!

drive escalation requests.

### SRAM_CTRL_EXEC_IF
The SRAM_CTRL IP has an interface to enable instruction execution from SRAM -
allowing the Icache to fetch instruction data from the SRAM for the CPU.
This interface contains the necessary Lifecycle and OTP structs to enable and
disable this functionality.

### UVM RAL Model
The SRAM_CTRL RAL model is created with the [`ralgen`]({{< relref "hw/dv/tools/ralgen/README.md" >}}) FuseSoC generator script automatically when the simulation is at the build stage.

It can be created manually by invoking [`regtool`]({{< relref "util/reggen/README.md" >}}):

<!-- TODO add description of second RAL model once implemented -->

### Reference models
The SRAM_CTRL testbench uses a SystemVerilog golden model of its address and data
[scrambling]({{< relref "hw/dv/sv/mem_bkdr_if/sram_scrambler_pkg.sv" >}}) to

> check that encryption and decryption are being performed correctly.

Reviewer 2: ```suggestion
check that encryption and decryption are performed correctly.
```

Author: fixed

#### Scoreboard
The `sram_ctrl_scoreboard` is primarily used for end to end checking.
It creates the following analysis ports to retrieve the data monitored by corresponding interface agents:
* tl_a_chan_fifo: TL address channel for CSR accesses
* tl_d_chan_fifo: TL data channel for CSR accesses
* alert_fifos: Alert handshakes
* sram_tl_a_chan_fifo: TL address channel for memory accesses
* sram_tl_d_chan_fifo: TL data channel for memory accesses

> * kdi_fifo: For key refresh operations from OTP

Reviewer: This is super nit: but I think we call it OTP_CTRL instead of OTP. (OTP can refer to the actual OTP memory)

Author: good point let me update this, thanks!

#### Scoreboard
The `sram_ctrl_scoreboard` is primarily used for end to end checking.
It creates the following analysis ports to retrieve the data monitored by corresponding interface agents:
* tl_a_chan_fifo: TL address channel for CSR accesses
* tl_d_chan_fifo: TL data channel for CSR accesses
* alert_fifos: Alert handshakes
* sram_tl_a_chan_fifo: TL address channel for memory accesses
* sram_tl_d_chan_fifo: TL data channel for memory accesses
* kdi_fifo: For key refresh operations from OTP
<!-- explain inputs monitored, flow of data and outputs checked -->

> All CSR accesses made to the SRAM_CTRL register file are tracked and predicted accordingly inside the scoreboard.

Reviewer 2: ```suggestion
All CSR accesses made to the SRAM_CTRL register file are tracked and predicted by the scoreboard.
```

Author: fixed, thanks!

> Verifying memory accesses is a little more involved of a process, and makes

Reviewer 2: ```suggestion
Verifying memory accesses is somewhat involved, and makes
```

Author: updated, thanks for the suggestion!

heavy use of the [`mem_bkdr_if`]({{< relref "hw/dv/sv/mem_bkdr_if/mem_bkdr_if.sv" >}}).

At a high level, whenever a TL access is made to memory, all relevant information is stored in a transaction-level struct

> (such as write_enable, address, write data, mask, key, and nonce).

Reviewer 2: It is unclear if these are all or some of the fields, or why these specific ones are mentioned.

Author: that's a good point - I'll add a brief description of the actual struct here.

Depending on the transaction type (read or write), we then hold the transaction
in limbo until it will be reflected in the underlying memory macro.
For example, all read transactions will return a response in 1 cycle, but a write may not be reflected in memory until many cycles later if it is immediately followed by a read request to the same address.
This data forwarding is modeled in the scoreboard such that we check all memory transactions

> in the exact same order that they are actually executed by the RAM primitive.

Reviewer 2: This could be rephrased. It seems there is a write buffer, and it can directly service reads if they match the write's address, so the read won't access the sram, and the write will have some extra delay.

Author: good idea - I rephrased this a bit to make it more clear, thanks!

> At this time, we use the memory backdoor interface to read the word in memory at the transaction's address

Reviewer 2: ```suggestion
When transactions complete, we use the memory backdoor interface to read the word in memory at the transaction's address
```

Author: fixed

> and then compare it against the requested write data or the read response data,

Reviewer 2: ```suggestion
and compare it against the requested write data or the read response data,
```

Author: done

> using an appropriate byte-mask depending on the granularity of the memory access.

Reviewer 2: ```suggestion
using a byte-mask corresponding to the granularity of the memory access.
```

Author: fixed

We use a specific set of methods in the memory backdoor interface that utilize
the memory scrambling golden model in order to provide the testbench with the
correct unencrypted data for these comparisons.

> For example, say we perform a memory access to address `0x0` to write a `1` into memory.

Reviewer 2: ```suggestion
For example, consider a memory access that writes `1` to address `0x0`.
```

Author: thanks for the suggestion - done!

Internally, the scrambling RAM primitive will map address `0x0` to some other address,
say scrambled address `0x4`, and will also encrypt the data.
In the scoreboard we detect this transaction on the TLUL bus and use the memory backdoor interface

> and scrambling golden model to perform a backdoor read to logical address `0x1`.

Reviewer 2: I don't understand: the requested access to address 0 is scrambled to address 4, and then you read from address 1? Could you clarify?

Author: whoops! this should be `0x0`, good catch.

> This will access the correct scrambled location in memory and decrypt the data there and return it to the scoreboard.

Reviewer 2: ```suggestion
This will access the correct scrambled location in memory, decrypt the data, and return it to the scoreboard.
```

Author: fixed, thanks

> In this way, it is easy for us to detect and track down any issues relating to incorrect address/data encryption/decryption.

Reviewer 2: ```suggestion
This makes it easier to diagnose issues relating to incorrect addresses, and to data encryption or decryption.
```

Author: that sounds much better, thanks!

The testbench will also sporadically trigger the DUT to issue a request for a new key and nonce from the OTP controller.
After receiving the fresh key and nonce from the KDI agent, those new values will then be used
for all encryption/decryption accesses using the memory backdoor interface.

If a lifecycle escalation request is issued during SRAM operation, the scoreboard will detect it
and then check that no further memory requests are accepted.
The testbench will then issue a system-level reset, as a lifecycle escalation is a terminal state.

To check that the SRAM's executable configurations are functioning correct is relatively more straightforward.

> First, it's important to note that all incoming TL memory transactions are all tagged as either

Reviewer 2: ```suggestion
First, it's important to note that all incoming TL memory transactions are tagged as either
```

Author: good catch - fixed
